### PR TITLE
feat: Add support for Interval

### PR DESCRIPTION
**Pull request by sagarwaal**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/nodejs-spanner/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> 🦕

### CODE REVIEW

File: system-test/spanner.ts [hunk 1]
```diff
@@ -6523,6 +6530,264 @@ describe('Spanner', () => {
...
+          const intervalQuery = (done, database, query, value) => {
+            database.run(query, (err, rows) => {
+              assert.ifError(err);
+              const queriedValue = rows[0][0].value;
+              assert.deepStrictEqual(queriedValue, value);
+              done();
+            });
+          };
+
+          it('GOOGLE_STANDARD_SQL should bind the value when param type interval is used', done => {
```
Comment (surbhigarg92) [inline system-test/spanner.ts:6551]: Please add a test for `GOOGLE_STANDARD_SQL should interval value is passed as string for both typed and untyped params.

var value = (new Interval(19, 768, BigInt('123456789123'))).toISO8601()

File: system-test/spanner.ts [hunk 2]
```diff
@@ -367,6 +368,7 @@ describe('Spanner', () => {
       googleSqlTable = DATABASE.table(TABLE_NAME);
       postgreSqlTable = PG_DATABASE.table(TABLE_NAME);
       if (IS_EMULATOR_ENABLED) {
+        // TODO: add columns using Interval Value and Interval Array Value.
```
Comment (surbhigarg92) [inline system-test/spanner.ts:371]: Why this is skipped ?

File: system-test/spanner.ts [hunk 3]
```diff
@@ -6523,6 +6530,264 @@ describe('Spanner', () => {
             });
           });
         });
+
+        describe('interval', () => {
```
Comment (surbhigarg92) [inline system-test/spanner.ts:6534]: Test cases for insert are missing. These test cases are only for query. Please add a block in describe('types')

File: test/codec.ts [hunk 4]
```diff
@@ -308,6 +308,515 @@ describe('codec', () => {
...
+            ),
+          },
+        ];
+
+        testCases.forEach(({input, expected}) => {
+          assert.deepStrictEqual(codec.Interval.fromISO8601(input), expected);
+        });
+      });
+
+      it('should throw error for invalid ISO8601 strings', () => {
```
Comment (surbhigarg92) [inline test/codec.ts:563]: Please also add test for null, undefined here and elsewhere

File: src/codec.ts [hunk 5]
```diff
@@ -394,6 +394,251 @@ export class PGOid extends WrappedNumber {
...
+
+  getDays(): number {
+    return this.days;
+  }
+
+  getNanoseconds(): bigint {
+    return this.nanoseconds;
+  }
+
+  static fromMonths(months: number): Interval {
```
Comment (surbhigarg92) [inline src/codec.ts:464]: Please add documentation for all public methods.

File: src/codec.ts [hunk 6]
```diff
@@ -394,6 +394,251 @@ export class PGOid extends WrappedNumber {
...
+ * @typedef Interval
+ * @see Spanner.interval
+ */
+export class Interval {
+  private months: number;
+  private days: number;
+  private nanoseconds: bigint;
+
+  private static readonly ISO8601_PATTERN: RegExp =
+    /^P(?!$)(-?\d+Y)?(-?\d+M)?(-?\d+D)?(T(?=-?[.,]?\d)(-?\d+H)?(-?\d+M)?(-?(((\d+)([.,]\d{1,9})?)|([.,]\d{1,9}))S)?)?$/;
```
Comment (surbhigarg92) [inline src/codec.ts:407]: Add code comments to explain the regex. Also elsewhere where its not self explanatory.

File: system-test/spanner.ts (hunk 1 above)
Comment (sagarwaal) [reply system-test/spanner.ts:6546]: done, added test for typed-param, skipped untyped param as library doesn't support untyped param.

File: src/codec.ts [hunk 7]
```diff
@@ -394,6 +394,251 @@ export class PGOid extends WrappedNumber {
...
+/**
+ * @typedef Interval
+ * @see Spanner.interval
+ */
+export class Interval {
+  private months: number;
+  private days: number;
+  private nanoseconds: bigint;
+
+  private static readonly ISO8601_PATTERN: RegExp =
```
[the remark this answers is not in the document]
Comment (sagarwaal) [reply src/codec.ts:409]: precise-date parsing will convert string to Date object, from that extracting original Interval is not possible. 

Interval is a duration without any fix start/end whereas Date object is a specified date.

File: system-test/spanner.ts (hunk 2 above)
Comment (sagarwaal) [reply system-test/spanner.ts]: Removed this comment, Interval is not a stored type, so we can't have columns of Interval type.

File: system-test/spanner.ts (hunk 3 above)
Comment (sagarwaal) [reply system-test/spanner.ts:6529]: This is intended,  Interval is not a stored type, so we can't have columns of Interval type.

File: test/codec.ts (hunk 4 above)
Comment (sagarwaal) [reply test/codec.ts:563]: Added tests for null and undefined.

File: src/codec.ts (hunk 5 above)
Comment (sagarwaal) [reply src/codec.ts:484]: added.

File: src/codec.ts (hunk 6 above)
Comment (sagarwaal) [reply src/codec.ts:410]: Added.

**Review by surbhigarg92 — APPROVED** [review]

**Review by rahul2393 — APPROVED** [review]
